src/stdlib_linalg_norms.fypp

```diff
@@ -0,0 +1,301 @@
+    ! Pure function interface, with order specification. On error, the code will stop
+    pure module function stdlib_linalg_norm_${rank}$D_order_${ii}$_${ri}$(a, order) result(nrm)
+        !> Input ${rank}$-d matrix a${ranksuffix(rank)}$
+        ${rt}$, intent(in) :: a${ranksuffix(rank)}$
+        !> Order of the matrix norm being computed.
+        ${it}$, intent(in) :: order
+        !> Norm of the matrix.
+        real(${rk}$) :: nrm
+                                    
+        call norm_${rank}$D_${ii}$_${ri}$(a, nrm=nrm, order=order)
+        
+    end function stdlib_linalg_norm_${rank}$D_order_${ii}$_${ri}$
+    
+    ! Function interface with output error
+    module function stdlib_linalg_norm_${rank}$D_order_err_${ii}$_${ri}$(a, order, err) result(nrm)
+        !> Input ${rank}$-d matrix a${ranksuffix(rank)}$
+        ${rt}$, intent(in) :: a${ranksuffix(rank)}$
+        !> Order of the matrix norm being computed.
+        ${it}$, intent(in) :: order
+        !> Output state return flag. 
+        type(linalg_state_type), intent(out) :: err                         
+        !> Norm of the matrix.
+        real(${rk}$) :: nrm                 
+                
+        call norm_${rank}$D_${ii}$_${ri}$(a, nrm=nrm, order=order, err=err)
+        
+    end function stdlib_linalg_norm_${rank}$D_order_err_${ii}$_${ri}$
+    
+    ! Internal implementation
+    pure module subroutine norm_${rank}$D_${ii}$_${ri}$(a, nrm, order, err)
+        !> Input ${rank}$-d matrix a${ranksuffix(rank)}$
+        ${rt}$, intent(in) :: a${ranksuffix(rank)}$
+        !> Norm of the matrix.
+        real(${rk}$), intent(out) :: nrm
+        !> Order of the matrix norm being computed.
+        ${it}$, intent(in) :: order
+        !> [optional] state return flag. On error if not requested, the code will stop
+        type(linalg_state_type), intent(out), optional :: err         
+        
+        type(linalg_state_type) :: err_
+        
+        integer(ilp) :: sze,norm_request
+        real(${rk}$) :: rorder
+        
+        sze = size(a,kind=ilp)
+        
+        ! Initialize norm to zero
+        nrm = 0.0_${rk}$
+        
+        ! Check matrix size
+        if (sze<=0) then
+            err_ = linalg_state_type(this,LINALG_VALUE_ERROR,'invalid matrix shape: a=',shape(a,kind=ilp))
+            call linalg_error_handling(err_,err)
+            return
+        end if
+        
+        ! Check norm request
+        call parse_norm_type(order,norm_request,err_)
+        if (err_%error()) then 
+            call linalg_error_handling(err_,err)
+            return
+        endif          
+        
+        select case(norm_request)
+            case(NORM_ONE)
+                nrm = sum( abs(a) )
+            case(NORM_TWO)            
+                #:if rt.startswith('complex')
+                nrm = sqrt( real( sum( a * conjg(a) ), ${rk}$) )
+                #:else
+                nrm = sqrt( sum( a ** 2 ) )
+                #:endif             
```

Comment:
What is the difference with the intrinsic `norm2`? Could we use it instead of `sqrt(sum(a**2))`?

Reply:
[437b96e](https://github.com/fortran-lang/stdlib/pull/871/commits/437b96eb020516d25048b810833884f69689ac8d)